### PR TITLE
feat: blog category

### PR DESCRIPTION
**Pull request by rugue**
# What does this PR do?
This PR Develops backend logic to securely handle adding blog categories by superadmins. The API allows superadmins to create new categories for organizing blog posts.

### Blog Category Creation:

* Implement server-side logic to add new blog categories.
* Restrict category creation capabilities to users with a super admin role.
* Authentication via token is required to access the category creation functionality.

### Validation and Sanitization:

* Validate and sanitize incoming category data to prevent injection attacks and ensure data integrity.

### Data Storage:

* Securely store new blog category details in the database, adhering to best practices for data security.

### Response Handling:

* Return appropriate success or error messages based on the outcome of the category creation process.
* Ensure error messages do not expose sensitive information and are handled gracefully.

### Purpose

Enhance backend functionality for adding blog categories securely, enabling superadmins to organize blog posts effectively with strict access control enforced via token authentication.

### Endpoint

Create Blog Category
* Description: Handles requests to create new blog categories.
* Endpoint: POST `/api/v1/blog-categories`

Request Body:

```JSON
{
    "name": "string"
}
```

Success Response:

- Header: `Status Code: 201 Created`
- Response Body:

```JSON
{
    "status": "success",
    "message": "Blog category created successfully.",
    "data": {
        "name": "Category Name"
    },
    "status_code": 201
}
```

Error Response:

- Status Code: `400 Bad Request`
	- Response Body:
```JSON
{
    "status": "error",
    "message": "Invalid request data. Please provide a valid category name.",
    "status_code": 400
}
```

Status Code: 401 Unauthorized
Response Body:

```JSON
{
    "status": "error",
    "message": "Unauthorized. Token is missing or invalid.",
    "status_code": 401
}
```

Status Code: 403 Forbidden (Role Issues)
Response Body:

```JSON
{
    "status": "error",
    "message": "Forbidden. You do not have permission to create blog categories.",
    "status_code": 403
}
```

Status Code: 500 Internal Server Error
Response Body:

```JSON
{
    "status": "error",
    "message": "Internal Server Error. Please try again later.",
    "status_code": 500
}
```

### Tasks

 * Implement a secure POST endpoint (/api/blog-categories) to handle requests for creating blog categories, requiring token authentication.
 * Validate and sanitize incoming data (category name) to prevent security vulnerabilities.
 * Securely store new blog category details in the database with role-based access control and token verification.
 * Handle and return appropriate success or error messages for category creation operations, ensuring token-based authentication and role restrictions are enforced.

### Testing

* Write unit tests to validate the functionality of the secure blog category creation endpoint, including successful creations and error handling for failed creations.
* Test token authentication and role-based access control mechanisms to ensure only superadmins with valid tokens can perform category creation operations.
* Perform integration testing to validate end-to-end functionality of the secure blog category creation process, focusing on data validation, storage, and authentication.

## Checklist
- [x] Implement secure POST endpoint (/api/v1/blog-categories) for creating blog categories with token authentication.
- [x] Validate and sanitize incoming category data to ensure security and integrity.
- [x] Securely store new blog category details in the database.
- [x] Restrict category creation to superadmins with role-based access control.
- [x] Handle and return appropriate success and error messages for category creation requests.
- [x] Write unit tests for the blog category creation endpoint.
- [x] Test token authentication and role-based access control mechanisms.
- [x] Perform integration testing to validate end-to-end functionality.

# Link to Issue
Link to this issue: [[FEAT]: Implement API Endpoint for Adding Blog Categories #142](https://github.com/hngprojects/hng_boilerplate_nestjs/issues/142)

### CODE REVIEW

File: src/modules/blog-category/blog-categories.controller.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+  @ApiResponse({ status: 201, description: 'Blog category created successfully.' })
+  @ApiResponse({ status: 400, description: 'Invalid request data. Please provide a valid category name.' })
+  @ApiResponse({ status: 401, description: 'Unauthorized. Token is missing or invalid.' })
+  @ApiResponse({ status: 403, description: 'Forbidden. You do not have permission to create blog categories.' })
+  @ApiResponse({ status: 500, description: 'Internal Server Error. Please try again later.' })
+  async create(@Body() createBlogCategoryDto: CreateBlogCategoryDto, @Request() req: { user: User }) {
+    const blogCategory = await this.blogCategoriesService.create(createBlogCategoryDto);
+    return {
+      status: 'success',
+      message: 'Blog category created successfully.',
```
Comment (Cobidex) [inline src/modules/blog-category/blog-categories.controller.ts:27]: Use constants for your messages, it's cleaner and scalable

File: src/modules/blog-category/dto/create-blog-category.dto.ts [hunk 2]
```diff
@@ -0,0 +1,9 @@
+import { ApiProperty } from '@nestjs/swagger';
+import { IsNotEmpty, IsString } from 'class-validator';
+
+export class CreateBlogCategoryDto {
+  @ApiProperty({ example: 'Technology' })
+  @IsNotEmpty({ message: 'Category name must not be empty.' })
```
Comment (Cobidex) [inline src/modules/blog-category/dto/create-blog-category.dto.ts:6]: No need for these messages, the validator will handle it for you.

**Review by Cobidex — APPROVED** [review]

**Review by Cobidex — DISMISSED** [review]

File: .husky/commit-msg [hunk 3]
```diff
@@ -1,4 +1,4 @@
 #!/bin/sh
```
Comment (CFCIfe) [inline .husky/commit-msg:1]: Why are we editing this file. It should be @Homoakin619 or someone else in DevOps responsibility

File: src/modules/blog-category/dto/create-blog-category.dto.ts [hunk 4]
```diff
@@ -0,0 +1,9 @@
+import { ApiProperty } from '@nestjs/swagger';
+import { IsNotEmpty, IsString } from 'class-validator';
+
+export class CreateBlogCategoryDto {
+  @ApiProperty({ example: 'Technology' })
+  @IsNotEmpty({})
+  @IsString({})
```
Comment (CFCIfe) [inline src/modules/blog-category/dto/create-blog-category.dto.ts:7]:
```suggestion
  @IsNotEmpty()
  @IsString()
```

File: src/modules/blog-category/blog-categories.controller.ts [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+  @Post()
+  @UseGuards(SuperAdminGuard)
+  @ApiBearerAuth()
+  @ApiOperation({ summary: 'Create a new blog category' })
+  @ApiResponse({ status: 201, description: 'Blog category created successfully.' })
+  @ApiResponse({ status: 400, description: 'Invalid request data. Please provide a valid category name.' })
+  @ApiResponse({ status: 401, description: 'Unauthorized. Token is missing or invalid.' })
+  @ApiResponse({ status: 403, description: 'Forbidden. You do not have permission to create blog categories.' })
+  @ApiResponse({ status: 500, description: 'Internal Server Error. Please try again later.' })
+  async create(@Body() createBlogCategoryDto: CreateBlogCategoryDto, @Request() req: { user: User }) {
```
Comment (CFCIfe) [inline src/modules/blog-category/blog-categories.controller.ts:23]: Remove `@Request() req: { user: User }` it's not being used
```suggestion
  async create(@Body() createBlogCategoryDto: CreateBlogCategoryDto) {
```

File: src/modules/blog-category/blog-categories.controller.ts [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    const blogCategory = await this.blogCategoriesService.create(createBlogCategoryDto);
+    return {
+      status: 'success',
+      message: 'Blog category created successfully.',
+      data: blogCategory,
+      status_code: 201,
+    };
```
Comment (CFCIfe) [inline src/modules/blog-category/blog-categories.controller.ts:30]: Move the return statement to your service file and like @Cobidex mentioned, use constants. check auth service for how this is done.

File: src/modules/blog-category/blog-categories.module.ts [hunk 7]
```diff
@@ -0,0 +1,14 @@
+import { Module } from '@nestjs/common';
+import { TypeOrmModule } from '@nestjs/typeorm';
+import { BlogCategory } from './entities/blog-category.entity';
+import { BlogCategoriesService } from './blog-categories.service';
+import { SuperAdminGuard } from 'src/guards/super-admin.guard';
```
Comment (CFCIfe) [inline src/modules/blog-category/blog-categories.module.ts:5]: The path used will fail if there is a change.

```suggestion
import { SuperAdminGuard } from '../../guards/super-admin.guard';
```

File: src/modules/blog-category/blog-categories.service.ts [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+
+@Injectable()
+export class BlogCategoriesService {
+  constructor(
+    @InjectRepository(BlogCategory)
+    private blogCategoryRepository: Repository<BlogCategory>
+  ) {}
+
+  async create(createBlogCategoryDto: CreateBlogCategoryDto): Promise<BlogCategory> {
+    const blogCategory = this.blogCategoryRepository.create(createBlogCategoryDto);
```
Comment (CFCIfe) [inline src/modules/blog-category/blog-categories.service.ts:15]:
```suggestion
  async createBlogCategory(createBlogCategoryDto: CreateBlogCategoryDto): Promise<BlogCategory> {
    const blogCategory = this.blogCategoryRepository.createBlogCategory(createBlogCategoryDto);
```

File: src/modules/blog-category/blog-categories.controller.ts [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+
+  @Post()
+  @UseGuards(SuperAdminGuard)
+  @ApiBearerAuth()
+  @ApiOperation({ summary: 'Create a new blog category' })
+  @ApiResponse({ status: 201, description: 'Blog category created successfully.' })
+  @ApiResponse({ status: 400, description: 'Invalid request data. Please provide a valid category name.' })
+  @ApiResponse({ status: 401, description: 'Unauthorized. Token is missing or invalid.' })
+  @ApiResponse({ status: 403, description: 'Forbidden. You do not have permission to create blog categories.' })
+  @ApiResponse({ status: 500, description: 'Internal Server Error. Please try again later.' })
```
Comment (Cobidex) [inline src/modules/blog-category/blog-categories.controller.ts:22]: Internal server error is not an error you expect your client to see in production. Please remove.

**Pull request closed without merging**